cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -0,0 +1,359 @@
+      super(String.format("NotificationSubscriberThread-%s", topic));
+      this.topic = topic;
+      this.messageId = messageId;
+      // TODO: [CDAP-11370] Need to be configured in cdap-default.xml. Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.readyJobs = new ArrayDeque<>();
+    }
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = processNotifications();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Fetch new notifications and update job queue
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long processNotifications() {
+      try {
+        final MessageFetcher fetcher = messagingContext.getMessageFetcher();
+        emptyFetchCount++;
+        transactional.execute(new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            fetchNotifications(context, fetcher);
+          }
+        });
+      } catch (Exception e) {
+        LOG.warn("Failed to get notification. Will retry in next run", e);
+        failureCount++;
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+      failureCount = 0;
+
+        try {
+          runReadyJobs();
+        } catch (Exception e) {
+          LOG.error("Failed to run scheduled programs", e);
+        }
+      // Back-off if it was empty fetch.
+      if (emptyFetchCount > 0) {
+        return 2000L;
+      }
+      return 0L; // No sleep if the fetch is non-empty
+    }
+
+    private void fetchNotifications(DatasetContext context, MessageFetcher fetcher) throws Exception {
+      try (CloseableIterator<Message> iterator = fetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                               topic, 100, messageId)) {
+        LOG.trace("Fetch with messageId = {}", messageId);
+        String currentMessageId = null;
+        while (iterator.hasNext() && isRunning()) {
+          emptyFetchCount = 0;
+          Message message = iterator.next();
+          Notification notification = null;
+          try {
+            notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
+                                         Notification.class);
+          } catch (JsonSyntaxException e) {
+            LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
```

Comment:
Better just update the `currentMessageId` and `continue`. Then you don't need to check if `notification == null` down there (which is also safer for future change).